### ISSUE TITLE
SIGPIPE when connection closed by peer

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#7133]
Despite the code to disable SIGPIPE, customers are still getting a SIGPIPE delivered to their applications. This is on CentOS:7, with libcurl 7.29.0, unclear what compiler or version of the library at this point.

**Comment by coryan** [comment]
I can only reproduce this if I call `std::signal(SIGPIPE, SIG_DFL);`.  However, that is not the recommended setting for multi-threaded applications using `libcurl`:

"When CURLOPT_NOSIGNAL is set to 1L, your application needs to deal with the risk of a SIGPIPE (that at least the OpenSSL backend can trigger). Note that setting CURLOPT_NOSIGNAL to 0L will not work in a threaded situation as there will be race where libcurl risks restoring the former signal handler while another thread should still ignore it."

https://curl.se/libcurl/c/threadsafe.html

And ignoring `SIGPIPE` is a recommended practice:

https://stackoverflow.com/questions/108183/how-to-prevent-sigpipes-or-handle-them-properly

My copy of Stevens' [Unix Network Programming](https://www.amazon.com/UNIX-Network-Programming-Networking-Sockets/dp/013490012X) suggests setting `SIGPIPE` to `SIG_IGN` too. After naming circumstances where one might use a signal handler for `SIGPIPE` it goes not to say:

> Be aware, however, that if multiple sockets are in use, the delivery of the signal does not tell us which socket encountered the error. If we need to know which `write` cause the error, then we must either ignore the signal or return from the signal handler and handle `EPIPE` from the `write`.
> "UNIX Network Programming, Volume 1, second edition, -- 5.13 `SIGPIPE` Signal, page 133

In summary, setting the signal handler to `SIGPIPE` to `SIG_DFL` is almost always a bug in the application.

**Comment by coryan** [comment]
Still waiting for a systematic way to repro this problem.
